### ISSUE TITLE
How to test TPS in orderer with kafka(the json file?)

### ISSUE THREAD

**Issue by shiningfrog** [hyperledger/caliper#299]
hello, how i can test performace of the fabric under kafka mode? how to fill the XX.json file in network/fabric-v1.2 ?

**Comment by houqinghui** [comment]
@shiningfrog  The `orderer.yaml` file is the configuration that is processed when the `orderer` node/binary is started. Caliper doesn't care about the topology of the ordering service, it just needs an orderer endpoint (currently only one endpoint is supported) to submit transactions to.

You can either modify the yaml file **before** starting the orderer node (not that simple with prebuilt docker images), or you can modify every setting available in that file through environment variables, which can be easily set in a `docker-compose` file.

For example, if the yaml has the setting `General.GenesisFile`, then you can insert the following line into the `environment` section of your compose file, under the orderer container:
`- ORDERER_GENERAL_GENESISFILE=/path/to/your/genesisfile`, like [here](https://github.com/hyperledger/caliper/blob/1fe8ba5bc3b5e9c9d9c5f0c905b2db487be3f1b7/network/fabric/2-org-1-peer/docker-compose-tls.yaml#L47).

Basically the rule is the following: for the orderer, prepend every environment variable with `ORDERER_`, and for every hierarchy in the setting, change the `.` to `_`, plus use capital letters.

**Comment by shiningfrog** [comment]
thank you. By the way, is there any plan to support multi endpoint in caliper design? @houqinghui

**Comment by aklenik** [comment]
@shiningfrog Pr #296 adds a new Fabric adapter, that supports multiple orderer (and of course peer) endpoints. It's not merged yet, so any feedback is welcome!

**Comment by quangtdn** [comment]
Hi there, I am trying to run Caliper in Distributed Clients (to test the Kafka mode if possible). 
According to the Caliper docs and my understanding, you first have to connect the client to the Zookeeper server and then run:
node benchmark/simple/main.js -c config-zookeeper.yaml -n XX-distributed-network.json

config-zookeeper.yaml is already provided in Caliper, but no blockchain network config file is provided to test the distributed clients. 
In case we don't specify XX-distributed-network.json, the program will use some XX.json in network/fabricv1.2/ directories to create a network instead of connecting to our Zookeeper network. 

Could you please provide instructions on how to test the distributed-clients mode, or a sample XX-distributed-network.json file? I know that testing distributed clients is still under the flag "experimental". Thank you very much!

**Comment by aklenik** [comment]
@quangtdn The distributed client mode of Caliper makes it possible to generate the workload from multiple machines. This is configured by the `config-zookeeper.yaml` file (or something similar to suit your needs). This config file defines the workload part of Caliper, which is dependent on Zookeeper in this case.

On the other hand, the other config file - the network descriptor - details the layout and endpoints of the backend blockchain network. It is independent of the workload generator mode of Caliper and that's why you won't find a zookeeper-specific network configuration file (not to be confused with a Zookeeper-based ordering service in Fabric, which is also transparent to Caliper, by the way). 

So use any example network config file you want that describes your backend network.

**Comment by quangtdn** [comment]
Thank you @aklenik . May I ask some more specific questions? 
1) I am a bit confused with the Zookeeper service here vs Zookeeper-based ordering service in Fabric. As I understand now, Zookeeper in Caliper is used only to connect the ZK clients and for the master node to distribute the workload to other ZK clients, and thus has nothing to do with Zookeeper-based Fabric. Please correct me if I am wrong. 

2) For example, now using 2 separate machines (namely PC1, PC2) I can bring up a network with 2 organizations, where each organization has 1 peers, and one solo orderer. This is actually the 2org1peer blockchain config in Caliper.
PC1 has: orderer, org1, org1.peer1
PC2 has: org2, org2.peer1

Now I want to test this network using 2 clients on PC3 and PC4 via Caliper. So these 2 clients should share the same config-zookeeper.yaml (modified by me) and 2org1peer.json . Then I can take the 2org1peer.json from Caliper with the following modifications:
- Remove the 2 commands that bring up and down the local fabric network.
- Change the port endpoints of the orderer, 2 orgs, and 2 peers to that of the above network. (Such variables that represent the endpoints are "url", "requests" (in peer), and "events" (in peer))
Is what I think here correct or sufficient?

Thank you much again. I am quite new to this and really have some confusion here.

**Comment by aklenik** [comment]
@quangtdn This will be a little long, but hopefully useful. 

1) [Zookeeper](https://zookeeper.apache.org/) - as a high-level service - provides a decentralized/group synchronization mechanism. It enables a crash fault tolerant operation for a service by employing multiple nodes. Fabric uses it in it's ordering service to avoid a single point of failure during the consensus (like the solo ordering service - the only node crashes and the network is unusable). Caliper uses Zookeeper for managing the (remote) client processes that generate the workload. Crash fault tolerance is not necessarily an important requirement here, Caliper could use any remote messaging/synchronization service. So these are different features of different projects relying on the same technology.

2) Caliper doesn't care about how your network is distributed. It only needs the endpoint addresses of your network (where the orderers, peers, and CAs are available). As long as a node is reachable through its address, you can deploy it anywhere. 

_General advice._
Node.js is a single threaded environment. You may have async patterns, but it's not really parallel, it's more like concurrent execution. If you want to generate a high-rate workload, you need to take this limitation into account. Caliper uses the configured rate controller to schedule transaction invocations. If you want to generate a 100 TPS workload with a single client (Node.js process), the rate controller needs to enable a transaction in every 10ms. Or every 1ms for a 1000 TPS workload. A single threaded execution can reach this limit rather quickly. Hence the following design.

Caliper spawns _multiple client processes_ that distribute the specified workload among each other. Since the execution of a transaction consists mostly of waiting for network responses (so it's not really a CPU intensive task), you can use (for example) 5-6 times as many local client processes as logical CPU cores you have. 

As a general advice, I wouldn't go above 100 TPS with a single client (but 70 TPS maybe a safer bet). So if you have a 16-core CPU, it should be able to handle about 50 clients, i.e., generating a 5000 TPS workload (but you should perform your own experiments/measurements). But you might reach the limit even of a strong CPU. That's when you use multiple machines to distribute the workload on the first level, then on each machine you use multiple local client processes to distribute the workload on a second (local) level.

My advice: you already have enough complications by testing a distributed Fabric network. Only use the distributed workload generation feature of Caliper, if you can't provide a single machine strong enough to handle the workload generation. Nowadays you can get pretty strong VMs at a good price from any IaaS providers that can handle high-rate workload generation. But if you want a _really_ high rate workload, then you will need multiple stronger VMs.

**Comment by quangtdn** [comment]
Hi @aklenik . Your clarifications and general advice are indeed really useful. It's also good to know the limit for generating workload in general, as I am doing some benchmarking on Hyperledger Fabric. Thank you!

**Comment by aklenik** [comment]
@shiningfrog Did you manage to run your tests successfully?

**Comment by shiningfrog** [comment]
hello, thanks, it worked. but if i can't repeat the test on the same chain. i have to rebuild a chian network. and the tps is about 20. i wonder if the transfer balance worked? how can i improve the tps? now all the docker files are allocated in one computer. @aklenik

**Comment by aklenik** [comment]
@shiningfrog Running Caliper with a local network is mainly for development and testing purposes (test chaincode, client module or new Caliper feature implementations). A production Fabric network will be distributed among multiple machines (belonging to multiple organizations). So if you want higher TPS, you need to experiment with your Fabric topology: the number of endorsers, ordering service configurations, etc. 

BTW, the Fabric adapter I suggested has been merged into the master branch, so you can now use the official Caliper repo :)

Since the Kafka orderer service issue is resolved, I'm closing this issue.
